samples/performance/bindless/bindless.cpp

```diff
@@ -0,0 +1,925 @@
+
+		VkDescriptorBufferInfo draw_command_descriptor = create_descriptor(*indirect_call_buffer);
+		VkWriteDescriptorSet   draw_command_write      = vkb::initializers::write_descriptor_set(_descriptor_set, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, 3, &draw_command_descriptor, 1);
+
+		VkDescriptorBufferInfo device_address_descriptor;
+		VkWriteDescriptorSet   device_address_write;
+		if (m_supports_buffer_device && device_address_buffer)
+		{
+			device_address_descriptor = create_descriptor(*device_address_buffer);
+			device_address_write      = vkb::initializers::write_descriptor_set(_descriptor_set, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, 4, &device_address_descriptor, 1);
+		}
+
+		std::vector<VkWriteDescriptorSet> write_descriptor_sets;
+		switch (target)
+		{
+			case Target::RenderPipeline:
+				write_descriptor_sets = {model_write, texture_array_write, scene_write, draw_command_write};
+				break;
+			case Target::ComputePipeline:
+				write_descriptor_sets = {model_write, scene_write, draw_command_write};
+				break;
+			case Target::AddressPipeline:
+				write_descriptor_sets = {model_write, scene_write, device_address_write};
+				break;
+		}
+
+		vkUpdateDescriptorSets(get_device().get_handle(), static_cast<uint32_t>(write_descriptor_sets.size()), write_descriptor_sets.data(), 0, VK_NULL_HANDLE);
+	};
+
+	// render pipeline
+	bind(descriptor_set, descriptor_set_layout, Target::RenderPipeline);
+
+	//  compute pipeline
+	bind(gpu_cull_descriptor_set, gpu_cull_descriptor_set_layout, Target::ComputePipeline);
+
+	// Device address pipeline
+	if (m_supports_buffer_device)
+	{
+		bind(device_address_descriptor_set, device_address_descriptor_set_layout, Target::AddressPipeline);
+	}
+}
+
+void BindlessResources::update_scene_uniform()
+{
+	if (!scene_uniform_buffer)
+	{
+		scene_uniform_buffer = std::make_unique<vkb::core::Buffer>(*device, sizeof(SceneUniform), VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT, VMA_MEMORY_USAGE_CPU_TO_GPU);
+	}
+	scene_uniform.proj = camera.matrices.perspective;
+	scene_uniform.view = camera.matrices.view;
+
+	scene_uniform_buffer->update(&scene_uniform, sizeof(scene_uniform), 0);
+}
+
+void BindlessResources::draw()
+{
+	ApiVulkanSample::prepare_frame();
+
+	// Command buffer to be submitted to the queue
+	submit_info.commandBufferCount = 1;
+	submit_info.pCommandBuffers    = &draw_cmd_buffers[current_buffer];
+
+	// Submit to queue
+	VK_CHECK(vkQueueSubmit(queue, 1, &submit_info, VK_NULL_HANDLE));
+
+	ApiVulkanSample::submit_frame();
+}
+
+void BindlessResources::render(float delta_time)
+{
+	if (!prepared)
+	{
+		return;
+	}
+
+	if (render_mode == GPU_DEVICE_ADDRESS && !m_supports_buffer_device)
+	{
+		render_mode = GPU;
+	}
+
+	if (m_requires_rebuild)
+	{
+		build_command_buffers();
+		m_requires_rebuild = false;
+	}
+
+	draw();
+
+	update_scene_uniform();
+
+	run_cull();
```

Comment:
The rendering done in `draw()` above reads the indirect arguments. Is there currently anything synchronizing the work in `run_cull()` which _writes_ the indirect arguments, i.e. preventing a WAR hazard? (There is a wait after the dispatch in `run_cull()`, preventing RAW hazard -- however I do wonder if there should be a pipeline barrier there to guarantee the written data is made available.)

Reply:
The synchronization should be corrected explicitly now.  Draw has the device wait idle (that I'd love to take out of the framework), the memory barrier and the fence in culling should be correctly sufficient; please double check.